### ISSUE TITLE
How to gather predict on ddp

### ISSUE THREAD

**Issue by MarsSu0618** [PyTorchLightning/pytorch-lightning#5257]
## Problem
I encountered some questions about ddp. Because I train mode with ddp on 2 gpus.
And when i test and predict test dataloader on `test_step()`, the predict result just half data be predicted.
ex:
```
original test data: 10000
predict result : 5000 (1/gpus)
```
How to solve it? use all_gather()? 
Hope someone can answer, I will thanks a lot.

**Comment by karlind** [comment]
You may have a look at `torch.distributed.all_gather`.

**Comment by karlind** [comment]
And for a specific usage,  have a look at [https://github.com/open-mmlab/mmdetection/blob/482f60fe55c364e50e4fc4b50893a25d8cc261b0/mmdet/apis/test.py#L160](https://github.com/open-mmlab/mmdetection/blob/482f60fe55c364e50e4fc4b50893a25d8cc261b0/mmdet/apis/test.py#L160)

**Comment by MarsSu0618** [comment]
@karlind my sample code follow as:
So i change `init_ids`, `attention_mask`, `token_type_ids` with `torch.distributed.all_gather`?

```python
def test_step(self, batch, batch_idx):
        init_ids = batch['init_ids']
        attention_mask = batch['attention_mask']
        token_type_ids = batch['token_type_ids']
        predictions = self.model(init_ids, attention_mask, token_type_ids)
```

**Comment by iamkucuk** [comment]
With DDP training, each GPU sees only their partition of the dataset, so each process can only evaluate a part of the dataset. You can use metrics package to automatic gather across all processes.

**Comment by MarsSu0618** [comment]
@iamkucuk 
How to do and integrate pytorch-lightning? Thanks.

**Comment by iamkucuk** [comment]
Hi @MarsSu0618
[Here is the documentation for the metrics package.](https://pytorch-lightning.readthedocs.io/en/latest/metrics.html#) You can use a existing metric, or create one for yourself. It does all the gathering for you.

**Comment by karlind** [comment]
Each process can predict part of the dataset, just predict as usual and gather all predicted results in `validation_epoch_end` or `test_epoch_end`. After that, evaluate with the whole results. You will get the exact performance.

**Comment by MarsSu0618** [comment]
@karlind 
Because i want to write predict result to file. So I also can use this method to process it?
And how to gather in this situation. use `all_gather()`? Thank a lot. it stucks..

**Comment by karlind** [comment]
Sure you can. [https://github.com/open-mmlab/mmdetection/blob/482f60fe55c364e50e4fc4b50893a25d8cc261b0/mmdet/apis/test.py#L160](https://github.com/open-mmlab/mmdetection/blob/482f60fe55c364e50e4fc4b50893a25d8cc261b0/mmdet/apis/test.py#L160) This function is a good example to collect results from multiple processes.

**Comment by awaelchli** [comment]
There is also a `LightningModule.write_predictions` method

**Comment by MarsSu0618** [comment]
@awaelchli 
But i haven't seen `LightningModule.write_predictions` how to use in document?

**Comment by awaelchli** [comment]
`self.write_prediction(name="name", value=some_tensor, filename='predictions.pt')`
it's not documented unfortunately.
You can then load the file later and inside is a dict with the name and value pairs. 


> It shows initializing ddp: GLOBAL_RANK: 0, MEMBER: 1/2, then always hangs. How to solve it?

same with ddp spawn? More details would be needed here. Make sure to run with the latest PL version.

**Comment by MarsSu0618** [comment]
@awaelchli 
Thanks for reply. But i want to check this method which can be use `csv` file?

Then, when i use `ddp spawn` still have the same problem and hang for a while. My pl version is `1.1.1`

**Comment by awaelchli** [comment]
> Thanks for reply. But i want to check this method which can be use csv file?

no it's not supported currently. you can load the pytorch dump and then write it to a csv.

> Then, when i use ddp spawn still have the same problem and hang for a while

But at which point did this occur? In your original message you wrote that you obtained some outputs so what changed since then? Some information how to reproduce would be needed here.

**Comment by Arij-Aladel** [comment]
I have the same problem using validation_epoch_end  I get half of results I do not need to save I need to do some further processing on (val_step_outputs) but I get half of results how to get all results from (val_step_outputs) then do what ever I want? I d

**Comment by Arij-Aladel** [comment]
> @karlind
> Because i want to write predict result to file. So I also can use this method to process it?
> And how to gather in this situation. use `all_gather()`? Thank a lot. it stucks..

@  MarsSu0618  Have you solved your problem? can yiu share the idea please?
